File: 733-flood-fill/733-flood-fill.js
/**
 * @param {number[][]} image
 * @param {number} sr
 * @param {number} sc
 * @param {number} newColor
 * @return {number[][]}
 */

const fill = function (image, sr, sc, oldColor, newColor) {
    if (sr < 0 || sc < 0 || sr >= image.length || sc >= image[0].length || image[sr][sc] !== oldColor) return;
    image[sr][sc] = newColor;
    fill(image, sr + 1, sc, oldColor, newColor);
    fill(image, sr - 1, sc, oldColor, newColor);
    fill(image, sr, sc + 1, oldColor, newColor);
    fill(image, sr, sc - 1, oldColor, newColor);
}

var floodFill = function(image, sr, sc, newColor) {
    const oldColor = image[sr][sc];
    
    if (oldColor === newColor) return image;
    
    fill(image, sr, sc, image[sr][sc], newColor);
    return image;
};

File: 733-flood-fill/733-flood-fill.rb
# @param {Integer[][]} image
# @param {Integer} sr
# @param {Integer} sc
# @param {Integer} color
# @return {Integer[][]}
def flood_fill(image, sr, sc, color)
    
    original_color = image[sr][sc]
    
    return image if original_color == color
    
    fill(image, sr, sc, original_color, color) 
    
    image
end


def fill(image, sr, sc, old_color, new_color)
    return unless sr.between?(0, image.length - 1) && sc.between?(0, image[0].length - 1) && image[sr][sc] == old_color
    
    image[sr][sc] = new_color
    fill(image, sr + 1, sc, old_color, new_color)   
    fill(image, sr - 1, sc, old_color, new_color)   
    fill(image, sr, sc + 1, old_color, new_color)   
    fill(image, sr, sc - 1, old_color, new_color)   
end